### PR TITLE
HADOOP-15563 S3Guard to create on-demand DDB tables

### PR DESCRIPTION
**Pull request by steveloughran**
HADOOP-15563 Full S3Guard Support for on-demand DDB tables

* If you create a table with read capacity == write capacity == 0, you get an on-demand table.
* change the defaults in Constants and core-defaults to be zero. You get an on-demand table by default.
* S3Guard docs reworked to cover the topic, including the updated defaults and why you should go to on-demand; also mentioned in the troubleshooting section.

It'd be nice if set-capacity could switch a table to on-demand, but the latest AWS SDK doesn't let you do that, even though the low level REST API does.

ITestS3GuardToolDynamoDB l
* ifecycle test creates an on-demand table and skips changing its capacity.
* Fix HADOOP-16187 ITestS3GuardToolDynamoDB test failures: test wasn't clearing enough per-bucket options.

ITestDynamoDBMetadataStore: review of every test case in so that
* Capacity is always set to on-demand before table creation
* Every test always cleans up its tables, even on assert failures.

Those tests have tended to leak tables in the past; these changes should eliminate it on uninterrupted test runs, and with on-demand capacity you don't get billed much until the next time you do an audit and delete of tables.

Also:

- removed @Unstable attribute from those Constants which I consider having been shipping too long to be changed.
- moved test-only S3GUARD_DDB_TEST_TABLE_NAME_KEY key to S3ATestConstants. It was tagged Unstable, and should be in the right file.

Change-Id: Ia44559354666db8027e574eab97983167ed930bf

Note: this patch includes HADOOP-16117 as precursor; that's a separate PR, (#818); its needed to be able to create on-demand tables

### CODE REVIEW

**Comment by steveloughran** [comment]
Test: S3 Ireland w/ s3guard + auth

File: hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/s3guard.md [hunk 1]
```diff
@@ -315,63 +322,90 @@ Next, you can choose whether or not the table will be automatically created
...
 
 <property>
   <name>fs.s3a.s3guard.ddb.table.capacity.write</name>
-  <value>100</value>
+  <value>0</value>
   <description>
     Provisioned throughput requirements for write operations in terms of
-    capacity units for the DynamoDB table.  Refer to related config
-    fs.s3a.s3guard.ddb.table.capacity.read before usage.
+    capacity units for the DynamoDB table.
```
Comment (hadoop-yetus) [inline hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/s3guard.md]: whitespace:end of line

File: hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/s3guard.md [hunk 2]
```diff
@@ -528,29 +568,31 @@ metadata store will be created with these tags in DynamoDB.
...
 Example 3
 
 ```bash
 hadoop s3guard init -meta dynamodb://ireland-team -tag tag1=first;tag2=second;
 ```
 
-Creates a table "ireland-team" with tags "first" and "second".
+Creates a table "ireland-team" with tags "first" and "second". The read and
+write capacity will be those of the site configuration's values of
+`fs.s3a.s3guard.ddb.table.capacity.read` and `fs.s3a.s3guard.ddb.table.capacity.write`;
```
Comment (hadoop-yetus) [inline hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/s3guard.md]: whitespace:end of line

File: hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/s3guard.md [hunk 3]
```diff
@@ -1015,20 +1061,33 @@ for S3Guard applications.
...
+can also be done explicitly in the `s3guard init` command by setting the
+read and write capacities to zero.
 
-*Enabling DynamoDB On-Demand for a S3Guard table*
 
-You cannot currently enable DynamoDB on-demand from the `s3guard` command
-when creating or updating a bucket.
+```bash
+hadoop s3guard init -meta dynamodb://ireland-team -write 0 -read 0 s3a://ireland-1
+```
```
Comment (hadoop-yetus) [inline hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/s3guard.md]: whitespace:end of line

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 33 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 5 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 152 | Maven dependency ordering for branch |
| +1 | mvninstall | 1281 | trunk passed |
| +1 | compile | 1027 | trunk passed |
| +1 | checkstyle | 134 | trunk passed |
| +1 | mvnsite | 879 | trunk passed |
| +1 | shadedclient | 1648 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 341 | trunk passed |
| 0 | spotbugs | 1733 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| 0 | findbugs | 18 | branch/hadoop-project no findbugs output file (findbugsXml.xml) |
| -1 | findbugs | 1727 | root in trunk has 6 extant findbugs warnings. |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 31 | Maven dependency ordering for patch |
| +1 | mvninstall | 1119 | the patch passed |
| +1 | compile | 976 | the patch passed |
| +1 | javac | 976 | the patch passed |
| -0 | checkstyle | 143 | root: The patch generated 7 new + 29 unchanged - 4 fixed = 36 total (was 33) |
| +1 | mvnsite | 916 | the patch passed |
| -1 | whitespace | 0 | The patch has 3 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply |
| +1 | xml | 2 | The patch has no ill-formed XML file. |
| +1 | shadedclient | 630 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 350 | the patch passed |
| 0 | findbugs | 15 | hadoop-project has no data from findbugs |
||| _ Other Tests _ |
| -1 | unit | 945 | root in the patch failed. |
| +1 | asflicense | 50 | The patch does not generate ASF License warnings. |
| | | 13746 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.util.TestDiskCheckerWithDiskIo |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-879/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/879 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient xml findbugs checkstyle |
| uname | Linux 427b5f3dbd2a 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 1ae062c |
| Default Java | 1.8.0_212 |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-879/1/artifact/out/branch-findbugs-root-warnings.html |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-879/1/artifact/out/diff-checkstyle-root.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-879/1/artifact/out/whitespace-eol.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-879/1/artifact/out/patch-unit-root.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-879/1/testReport/ |
| Max. process+thread count | 1420 (vs. ulimit of 5500) |
| modules | C: hadoop-project hadoop-common-project/hadoop-common hadoop-tools/hadoop-aws . U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-879/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3GuardTool.java [hunk 4]
```diff
@@ -434,7 +434,8 @@ public abstract int run(String[] args, PrintStream out) throws Exception,
         "\n" +
         "  URLs for Amazon DynamoDB are of the form dynamodb://TABLE_NAME.\n" +
         "  Specifying both the -" + REGION_FLAG + " option and an S3A path\n" +
-        "  is not supported.";
+        "  is not supported.\n"
+        + "To create a table with per-request billing, set the read and write capaciies to 0";
```
Comment (bgaborg) [inline hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3GuardTool.java]: typo: capaciies -> capacities

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/DDBCapacities.java [hunk 5]
```diff
@@ -82,7 +79,7 @@ public String toString() {
   }
 
   /**
-   * Is the the capacity that of a pay-on-demand table?
+   * Is the the capacity that of a pay-per-request table?
```
Comment (bgaborg) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/DDBCapacities.java]: I don't understand this change. What's the diff between on-demand and per-request?

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/ITestDynamoDBMetadataStore.java [hunk 6]
```diff
@@ -274,7 +300,9 @@ public void testInitialize() throws IOException {
         getTestTableName("testInitialize");
     final Configuration conf = s3afs.getConf();
     conf.set(S3GUARD_DDB_TABLE_NAME_KEY, tableName);
-    try (DynamoDBMetadataStore ddbms = new DynamoDBMetadataStore()) {
+    enableOnDemand(conf);
```
Comment (bgaborg) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/ITestDynamoDBMetadataStore.java]: we could add this to @Before, right?

Comment (bgaborg) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/ITestDynamoDBMetadataStore.java]: You add this in @BeforeClass as well. Is there a point to add this to all tests as well? Maybe add this to `@Before` only?

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/Constants.java [hunk 7]
```diff
@@ -445,7 +445,6 @@ private Constants() {
    * This config has no default value. If the user does not set this, the
    * S3Guard will operate table in the associated S3 bucket region.
    */
-  @InterfaceStability.Unstable
```
[the remark this answers is not in the document]
Comment (steveloughran) [reply hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/Constants.java]: yeah, its time

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/Constants.java [hunk 8]
```diff
@@ -465,36 +463,45 @@ private Constants() {
    * For example:
    * fs.s3a.s3guard.ddb.table.tag.mytag
    */
-  @InterfaceStability.Unstable
   public static final String S3GUARD_DDB_TABLE_TAG =
       "fs.s3a.s3guard.ddb.table.tag.";
 
-  /**
```
[the remark this answers is not in the document]
Comment (steveloughran) [reply hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/Constants.java]: +1. I'm thinking we also need to split internal from public

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3GuardTool.java (hunk 4 above)
Comment (steveloughran) [reply hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3GuardTool.java]: fixed; also splitting the line in both the IDE and in the printed text

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/DDBCapacities.java (hunk 5 above)
Comment (steveloughran) [reply hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/DDBCapacities.java]: The AWS docs say "on demand" or "pay by request"; I was trying to be consistent but gave up. This is a leftover from an attempt/revert of actually changing the method name. Changing to "on demand" (and not pay-on-demand)

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/ITestDynamoDBMetadataStore.java (hunk 6 above)
Comment (steveloughran) [reply hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/ITestDynamoDBMetadataStore.java]: it's done in the beforeClass for the static table; for the others its done to to the config which comes from the FS.
I'm going to change it to set it in the FS config before we set the FS. If someone has per-bucket options to set the capacity then things will be different, but I'd be surprised if anyone ever does that

Comment (steveloughran) [reply hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/ITestDynamoDBMetadataStore.java]: the beforeClass is for the static ddbms; the others were for the new ones

Comment (steveloughran) [reply hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/ITestDynamoDBMetadataStore.java]: I'll be doing it in the FS configuration and the one for the static metastore -they are different configurations

**Comment by steveloughran** [comment]
BTW, HADOOP-15183 will be fixing the ITestDynamoDBMetadataStoreScale tests to move from verifying that DDB throttling is recovered from to becoming tests of the scalability of how our client uses DDB. Which will remove the current assume(!ondemand) change of HADOOP-16118. Not relevant for this patch except to explain why I'm not doing anything with that test in this one

**Comment by steveloughran** [comment]
Update: we do need the per-test setting of the capacity, else your created tables have numbers. Assumption: FS caching may complicate things: setting the capacity on the new FS doesn't get picked up if another FS is used. 
Factoring out the getFS.getConf() + patch routine, and also calling clone() so that each tests' changes are isolated from the others. Without that (as we are today) one test could interfere with its successor.

**Comment by steveloughran** [comment]
Latest PR is on trunk and the updated SDK; removed the SDK part of this patch

Tested: S3 Ireland with `-Dparallel-tests -DtestsThreadCount=8 -Ds3guard -Ddynamo`

No tests failed. Not even a little one :)

File: hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/s3guard.md [hunk 9]
```diff
@@ -315,63 +322,90 @@ Next, you can choose whether or not the table will be automatically created
...
+
+Next, you need to decide whether to use On-Demand DynamoDB and its
+pay-per-request billing (recommended), or to explicitly request a
+provisioned IO capacity.
 
-Next, you need to set the DynamoDB read and write throughput requirements you
-expect to need for your cluster.  Setting higher values will cost you more
-money.  *Note* that these settings only affect table creation when
+Before AWS offered pay-per-request billing, the sole billing mechanism,
+was "provisioned capacity". This mechanism requires you to choose
```
Comment (hadoop-yetus) [inline hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/s3guard.md]: whitespace:end of line

File: hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/s3guard.md [hunk 10]
```diff
@@ -315,63 +322,90 @@ Next, you can choose whether or not the table will be automatically created
...
+Larger I/O capacities cost more.
+
+With the introduction of On-Demand DynamoDB, you can now avoid paying for
+provisioned capacity by creating an on-demand table.
+With an on-demand table you are not throttled if your DynamoDB requests exceed
+any pre-provisioned limit, nor do you pay per hour even when a table is idle.
+
+You do, however, pay more per DynamoDB operation.
+Even so, the ability to cope with sudden bursts of read or write requests, combined
+with the elimination of charges for idle tables, suit the use patterns made of
```
Comment (hadoop-yetus) [inline hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/s3guard.md]: whitespace:end of line

File: hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/s3guard.md [hunk 11]
```diff
@@ -512,7 +546,13 @@ hadoop s3guard init -meta URI ( -region REGION | s3a://BUCKET )
 Creates and initializes an empty metadata store.
 
 A DynamoDB metadata store can be initialized with additional parameters
-pertaining to [Provisioned Throughput](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.ProvisionedThroughput.html):
+pertaining to capacity.
```
Comment (hadoop-yetus) [inline hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/s3guard.md]: whitespace:end of line

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 47 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 6 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 69 | Maven dependency ordering for branch |
| +1 | mvninstall | 1148 | trunk passed |
| +1 | compile | 1042 | trunk passed |
| +1 | checkstyle | 145 | trunk passed |
| +1 | mvnsite | 121 | trunk passed |
| +1 | shadedclient | 1046 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 93 | trunk passed |
| 0 | spotbugs | 67 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 191 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 23 | Maven dependency ordering for patch |
| +1 | mvninstall | 80 | the patch passed |
| +1 | compile | 1008 | the patch passed |
| +1 | javac | 1008 | the patch passed |
| -0 | checkstyle | 150 | root: The patch generated 6 new + 31 unchanged - 5 fixed = 37 total (was 36) |
| +1 | mvnsite | 124 | the patch passed |
| -1 | whitespace | 0 | The patch has 3 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply |
| +1 | xml | 1 | The patch has no ill-formed XML file. |
| +1 | shadedclient | 736 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 94 | the patch passed |
| +1 | findbugs | 206 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 563 | hadoop-common in the patch passed. |
| +1 | unit | 292 | hadoop-aws in the patch passed. |
| +1 | asflicense | 46 | The patch does not generate ASF License warnings. |
| | | 7230 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-879/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/879 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient xml findbugs checkstyle |
| uname | Linux 4d76f68c8ed7 4.4.0-144-generic #170~14.04.1-Ubuntu SMP Mon Mar 18 15:02:05 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / f7c77b3 |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-879/4/artifact/out/diff-checkstyle-root.txt |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-879/4/artifact/out/whitespace-eol.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-879/4/testReport/ |
| Max. process+thread count | 1714 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-tools/hadoop-aws U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-879/4/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/s3guard.md (hunk 9 above)
Comment (hadoop-yetus) [inline hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/s3guard.md]: whitespace:end of line

File: hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/s3guard.md (hunk 10 above)
Comment (hadoop-yetus) [inline hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/s3guard.md]: whitespace:end of line

File: hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/s3guard.md (hunk 11 above)
Comment (hadoop-yetus) [inline hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/s3guard.md]: whitespace:end of line

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 31 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 6 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 41 | Maven dependency ordering for branch |
| +1 | mvninstall | 1114 | trunk passed |
| +1 | compile | 1121 | trunk passed |
| +1 | checkstyle | 145 | trunk passed |
| +1 | mvnsite | 122 | trunk passed |
| +1 | shadedclient | 1045 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 95 | trunk passed |
| 0 | spotbugs | 67 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 194 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 23 | Maven dependency ordering for patch |
| +1 | mvninstall | 82 | the patch passed |
| +1 | compile | 1087 | the patch passed |
| +1 | javac | 1087 | the patch passed |
| +1 | checkstyle | 170 | root: The patch generated 0 new + 31 unchanged - 5 fixed = 31 total (was 36) |
| +1 | mvnsite | 138 | the patch passed |
| -1 | whitespace | 0 | The patch has 3 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply |
| +1 | xml | 1 | The patch has no ill-formed XML file. |
| +1 | shadedclient | 747 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 105 | the patch passed |
| +1 | findbugs | 244 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 616 | hadoop-common in the patch passed. |
| +1 | unit | 299 | hadoop-aws in the patch passed. |
| +1 | asflicense | 50 | The patch does not generate ASF License warnings. |
| | | 7476 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-879/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/879 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient xml findbugs checkstyle |
| uname | Linux 8f02a7b4c84e 4.4.0-143-generic #169~14.04.2-Ubuntu SMP Wed Feb 13 15:00:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 1f244b4 |
| Default Java | 1.8.0_212 |
| whitespace | https://builds.apache.org/job/hadoop-multibranch/job/PR-879/5/artifact/out/whitespace-eol.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-879/5/testReport/ |
| Max. process+thread count | 1322 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common hadoop-tools/hadoop-aws U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-879/5/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by mackrorysd** [comment]
+1. I can fix the whitespace as I commit.

**Comment by steveloughran** [comment]
Thanks, I can do the merge; I fix it when I do it on my desktop (That is when I Don't also break the build, pull in the wrong artifacts, etc, etc)

**Comment by steveloughran** [comment]
applied to trunk. Thanks for the review. This is a nice feature. Pity about the fact we can't yet change the mode of an existing table: I've had to go through the console and manually switch every single DDB table over in our test account.

**Comment by steveloughran** [comment]
Last little refactoring caused whitespace issues. Will fix
```
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/InternalConstants.java:24:public class InternalConstants {:1: Utility classes should not have a public or default constructor. [HideUtilityClassConstructor]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/RenameOperation.java:153:  public RenameOperation(:10: More than 7 parameters (found 8). [ParameterNumber]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/RenameOperation.java:399:      final S3ALocatedFileStatus sourceStatus,:34: 'sourceStatus' hides a field. [HiddenField]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/RenameOperation.java:438:  private Path copySourceAndUpdateTracker(:16: More than 7 parameters (found 8). [ParameterNumber]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/RenameOperation.java:439:      final RenameTracker renameTracker,:27: 'renameTracker' hides a field. [HiddenField]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/RenameOperation.java:440:      final Path sourcePath,:18: 'sourcePath' hides a field. [HiddenField]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/RenameOperation.java:444:      final Path destPath,:18: 'destPath' hides a field. [HiddenField]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/RenameOperation.java:445:      final String destKey,:20: 'destKey' hides a field. [HiddenField]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/RenameOperation.java:487:      final RenameTracker renameTracker,:27: 'renameTracker' hides a field. [HiddenField]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/RenameOperation.java:488:      final List<DeleteObjectsRequest.KeyVersion> keysToDelete,:51: 'keysToDelete' hides a field. [HiddenField]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/RenameOperation.java:489:      final List<Path> pathsToDelete):24: 'pathsToDelete' hides a field. [HiddenField]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/RenameOperation.java:541:        final Path path,:9: Redundant 'final' modifier. [RedundantModifier]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/RenameOperation.java:542:        final String eTag,:9: Redundant 'final' modifier. [RedundantModifier]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/RenameOperation.java:543:        final String versionId,:9: Redundant 'final' modifier. [RedundantModifier]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/RenameOperation.java:544:        final long len);:9: Redundant 'final' modifier. [RedundantModifier]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/RenameOperation.java:552:        final S3AFileStatus fileStatus);:9: Redundant 'final' modifier. [RedundantModifier]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/RenameOperation.java:561:        final FileStatus fileStatus);:9: Redundant 'final' modifier. [RedundantModifier]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/RenameOperation.java:563:    /**: First sentence should end with a period. [JavadocStyle]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/RenameOperation.java:633:        final List<DeleteObjectsRequest.KeyVersion> keysToDelete,:9: Redundant 'final' modifier. [RedundantModifier]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/RenameOperation.java:634:        final boolean deleteFakeDir,:9: Redundant 'final' modifier. [RedundantModifier]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/RenameOperation.java:635:        final List<Path> undeletedObjectsOnFailure):9: Redundant 'final' modifier. [RedundantModifier]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/StoreContext.java:119:  public StoreContext(:10: More than 7 parameters (found 17). [ParameterNumber]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java:50:import org.apache.hadoop.fs.s3a.Tristate;:8: Unused import - org.apache.hadoop.fs.s3a.Tristate. [UnusedImports]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java:525:   * {@link MetadataStore#addAncestors(Path, ITtlTimeProvider, BulkOperationState)}.: Line is longer than 80 characters (found 84). [LineLength]
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3GuardTool.java:1163:        clearBucketOption(unguardedConf, fsURI.getHost(), S3_METADATA_STORE_IMPL);: Line is longer than 80 characters (found 82). [LineLength]
./hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/commit/terasort/AbstractCommitTerasortIT.java:243:  public void test_200_directory_deletion() throws Throwable {:15: Name 'test_200_directory_deletion' must match pattern '^[a-z][a-zA-Z0-9]*$'. [MethodName]
./hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/impl/TestPartialDeleteFailures.java:74:  private static final ContextAccessors contextAccessors:41: Name 'contextAccessors' must match pattern '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'. [ConstantName]
./hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/impl/TestPartialDeleteFailures.java:349:    public BulkOperationState initiateBulkWrite(final BulkOperationState.OperationType operation,: Line is longer than 80 characters (found 97). [LineLength]
./hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/ITestDynamoDBMetadataStore.java:988:          () -> ddbms.prune(PruneMode.ALL_BY_MODTIME,0));:53: ',' is not followed by whitespace. [WhitespaceAfter]
./hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/ITestDynamoDBMetadataStore.java:1169:        null );:14: ')' is preceded with whitespace. [ParenPad]
```

**Pull request closed without merging**